File: C++/CircularQueue.java
#include <iostream>
using namespace std;

class circularQueue {
public:
    int front, rear, maxSize;
    int* queue;

    circularQueue(int size) {
        maxSize = size;
        front = -1;
        rear = -1;
        queue = new int[maxSize];
    }

    ~circularQueue() {
        delete[] queue;
    }

    void enqueue(int value) {
        if (((rear == maxSize - 1) && (front == 0)) || (front == rear - 1)) {
            cout << "queue overflow" << endl;
        } else {
            if (front == -1 && rear == -1) {
                front = 0;
                rear = 0;
            } else if (front != 0 && rear == maxSize - 1) {
                rear = 0;
            } else {
                rear++;
            }
        }
        queue[rear] = value;
        cout << "inserted: " << value << endl;
    }

    void dequeue() {
        if (front == -1 && rear == -1) {
            cout << "stack underflow" << endl;
        } else {
            if (front == rear) {
                front = -1;
                rear = -1;
            } else if (front == maxSize - 1 && front != rear) {
                front = 0;
            } else {
                front++;
            }
        }
    }

    void display() {
        if (front <= rear) {
            for (int i = front; i <= rear; i++) {
                cout << queue[i] << " ";
            }
        } else {
            for (int i = front; i < maxSize; i++) {
                cout << queue[i] << " ";
            }
            for (int i = 0; i <= rear; i++) {
                cout << queue[i] << " ";
            }
        }
        cout << endl;
    }

    bool isEmpty() {
        if (front == -1 && rear == -1) {
            return true;
        } else return false;
    }

    bool isFull() {
        if (((rear == maxSize - 1) && (front == 0)) || (front == rear - 1)) {
            return true;
        } else return false;
    }
};

int main() {
    int size;
    cout << "enter size of circular queue: " << endl;
    cin >> size;

    circularQueue var(size);

    while (true) {
        cout << "Enter a choice: " << endl;
        cout << "1.Enqueue; 2.Dequeue; 3.Display; 4.isEmpty; 5.isFull; 6.Exit" << endl;

        int choice;
        cin >> choice;

        switch (choice) {
            case 1: {
                cout << "---ENQUEUE FUNCTION---" << endl;
                cout << "Enter value to enqueue: " << endl;
                int val;
                cin >> val;
                var.enqueue(val);
                break;
            }
            case 2:
                cout << "---DEQUEUE FUNCTION---" << endl;
                var.dequeue();
                break;

            case 3:
                cout << "---DISPLAY FUNCTION---" << endl;
                var.display();
                break;

            case 4:
                cout << "---IS EMPTY FUNCTION---" << endl;
                cout << (var.isEmpty() ? "true" : "false") << endl;
                break;

            case 5:
                cout << "---IS FULL FUNCTION---" << endl;
                cout << (var.isFull() ? "true" : "false") << endl;
                break;

            case 6:
                cout << "Exiting..." << endl;
                return 0;

            default:
                cout << "Invalid choice!" << endl;
        }
    }

    return 0;
}

File: Python/CircularQueue.py
class CircularQueue:
    def __init__(self, size):
        self.max_size = size
        self.front = -1
        self.rear = -1
        self.queue = [None] * size

    def enqueue(self, value):
        if ((self.rear == self.max_size - 1) and (self.front == 0)) or (self.front == self.rear - 1):
            print("Queue overflow")
        else:
            if self.front == -1 and self.rear == -1:
                self.front = 0
                self.rear = 0
            elif self.front != 0 and self.rear == self.max_size - 1:
                self.rear = 0
            else:
                self.rear += 1
            self.queue[self.rear] = value
            print(f"Inserted: {value}")

    def dequeue(self):
        if self.front == -1 and self.rear == -1:
            print("Queue underflow")
        else:
            if self.front == self.rear:
                self.front = -1
                self.rear = -1
            elif self.front == self.max_size - 1 and self.front != self.rear:
                self.front = 0
            else:
                self.front += 1

    def display(self):
        if self.front == -1 and self.rear == -1:
            print("Queue is empty")
            return
        if self.front <= self.rear:
            print(" ".join(str(self.queue[i]) for i in range(self.front, self.rear + 1)))
        else:
            part1 = [str(self.queue[i]) for i in range(self.front, self.max_size)]
            part2 = [str(self.queue[i]) for i in range(0, self.rear + 1)]
            print(" ".join(part1 + part2))

    def is_empty(self):
        return self.front == -1 and self.rear == -1

    def is_full(self):
        return ((self.rear == self.max_size - 1) and (self.front == 0)) or (self.front == self.rear - 1)


def main():
    size = int(input("Enter size of circular queue: "))
    cq = CircularQueue(size)

    while True:
        print("\n1.Enqueue  2.Dequeue  3.Display  4.isEmpty  5.isFull  6.Exit")
        choice = int(input("Enter a choice: "))

        if choice == 1:
            print("---ENQUEUE FUNCTION---")
            val = int(input("Enter value to enqueue: "))
            cq.enqueue(val)
        elif choice == 2:
            print("---DEQUEUE FUNCTION---")
            cq.dequeue()
        elif choice == 3:
            print("---DISPLAY FUNCTION---")
            cq.display()
        elif choice == 4:
            print("---IS EMPTY FUNCTION---")
            print(cq.is_empty())
        elif choice == 5:
            print("---IS FULL FUNCTION---")
            print(cq.is_full())
        elif choice == 6:
            print("Exiting...")
            break
        else:
            print("Invalid choice!")

if __name__ == "__main__":
    main()